component/src/org/wso2/ballerina/connectors/oauth2/ClientConnector.bal

```diff
@@ -0,0 +1,173 @@
+package org.wso2.ballerina.connectors.oauth2;
+
+import ballerina.net.http;
+
+@Description { value:"OAuth2 client connector"}
+@Param { value:"baseUrl: The endpoint base url"}
+@Param { value:"accessToken: The access token of the account"}
+@Param { value:"clientId: The client Id of the account"}
+@Param { value:"clientSecret: The client secret of the account"}
+@Param { value:"refreshToken: The refresh token of the account"}
+@Param { value:"refreshTokenEP: The refresh token endpoint url"}
+public connector ClientConnector (string baseUrl, string accessToken, string clientId, string clientSecret,
+                                  string refreshToken, string refreshTokenEP, string refreshTokenPath) {
+
+    endpoint<http:HttpClient> httpConnectorEP {
+        create http:HttpClient(baseUrl, {});
+    }
+
+    string accessTokenValue;
+    http:HttpConnectorError e;
+
+    @Description { value:"Get with OAuth2 authentication"}
+    @Param { value:"path: The endpoint path"}
+    @Param { value:"request: The request of the method"}
+    @Return { value:"response object"}
+    @Return { value:"Error occured during HTTP client invocation." }
+    action get (string path, http:Request request) (http:Response, http:HttpConnectorError) {
+        http:Response response = {};
+
+        accessTokenValue = constructAuthHeader (request, accessTokenValue, accessToken);
+        response, e = httpConnectorEP.get (path, request);
+
+        if ((response.getStatusCode() == 401) && (refreshToken != "" || refreshToken != "null")) {
+            request = {};
+            accessTokenValue = getAccessTokenFromRefreshToken(request, accessToken, clientId, clientSecret, refreshToken,
+                                                              refreshTokenEP, refreshTokenPath);
+            response, e = httpConnectorEP.get (path, request);
+        }
+
+        return response, e;
+    }
+
+    @Description { value:"Post with OAuth2 authentication"}
+    @Param { value:"path: The endpoint path"}
+    @Param { value:"request: The request of the method"}
+    @Return { value:"response object"}
+    @Return { value:"Error occured during HTTP client invocation." }
+    action post (string path, http:Request originalRequest) (http:Response, http:HttpConnectorError) {
+        http:Response response = {};
+        json originalPayload = originalRequest.getJsonPayload();
+
+        accessTokenValue = constructAuthHeader (originalRequest, accessTokenValue, accessToken);
+        response, e = httpConnectorEP.post (path, originalRequest);
+
+        http:Request request = {};
+        request.setJsonPayload(originalPayload);
+
+        if ((response.getStatusCode() == 401) && (refreshToken != "" || refreshToken != "null")) {
```

Comment:
These two lines are repeated in all the operations. That is a sign that it can be used as a common code to be used by several places. These two lines can be wrapped in to a boolean returning function like "checkAndRefreshToken", where it will check if the response is a 401, and then generates a new access token if so and returns it. We can either do a multi-return with a boolean and a string, or else, just return a string and check if it's empty/null. (null is supported in v0.95.3).